parser-typechecker/src/Unison/Typechecker/Context.hs

```diff
@@ -1369,6 +1376,35 @@ ungeneralize' (Type.Forall' t) = do
   first (v:) <$> ungeneralize' t
 ungeneralize' t = pure ([], t)
 
+skolemize
+  :: Var v
+  => Ord loc
+  => (Type v loc -> Set (TypeVar v loc))
+  -> Type v loc
+  -> M v loc (Type v loc)
+skolemize forced (Type.ForallsNamed' vs ty) = do
+  urn <- for uvs $ \u -> (,) u <$> freshenTypeVar u
+  srn <- for svs $ \u -> (,) u <$> freshenTypeVar u
+  let uctx = existential . snd <$> urn
+      sctx = Universal . snd <$> srn
+      rn = (fmap (existential' () B.Blank) <$> urn)
+        ++ (fmap (universal' ()) <$> srn)
+  appendContext $ uctx ++ sctx
+  pure $ foldl (flip $ uncurry ABT.substInheritAnnotation) ty rn
+  where
+  fovs = forced ty
+  (uvs, svs) = partition (`Set.member` fovs) vs
+skolemize _ ty = pure ty
+
+forcedEffect :: Type v loc -> Set (TypeVar v loc)
+forcedEffect (Type.Arrow' _ o) = forcedEffect o
+forcedEffect (Type.Effect1' es _) = Type.freeVars es
```

Comment:
I was thinking this wasn't quite right if arrows other than the final one have effects, but that shouldn't be a valid ability declaration (though we don't enforce that).

Reply:
Is that because `X ->{E} Y ->{F} Z` is represented like `Arrow X (Effect E (Arrow Y (Effect F Z)))`?

Reply:
Right

Reply:
So, I was thinking about this randomly. My question is: how is this actually supposed to work? `X ->{E} Y ->{F} Z` is *not* the type of an `F` effect. It is of an `E` effect that produces a function with an `F` effect. If it is being put in the `F` ability, then that is wrong. The type checker might not prevent someone from writing it that way, but that means it is allowing something that doesn't make sense, I think.

Reply:
Right. It doesn't make sense. The typechecker should probably give a nice error for this.